### PR TITLE
Op registration: indexing

### PR DESCRIPTION
**Pull request by asi1024**
Part of #6692.

### CODE REVIEW

File: chainerx_cc/chainerx/routines/indexing.h [hunk 1]
```diff
@@ -7,22 +7,59 @@
...
 
+// Adds each slice of `b` along the axis `axis` to `a`'s corresponding slices, specified by `indices`.
+// The result is assigned in `out. Input arrays `a`, `indices`, and `b` are not altered.
+//
+// TODO(niboshi): This function may be replaced with full-featured assignable advanced indexing.
+//
+// `axis` must be within [0, b.ndim()).
+// `indices` must have dtype kind of either kInt or kUInt.
+
+class AddAtOp : public Op {
```
[the remark this answers is not in the document]
Comment (asi1024) [reply chainerx_cc/chainerx/routines/indexing.h]: Fixed.

File: chainerx_cc/chainerx/routines/indexing.cc [hunk 2]
```diff
@@ -173,15 +165,17 @@ Array Take(const Array& a, const Array& indices, int8_t axis) {
...
 
     BackwardBuilder bb{"take", a, out};
     if (BackwardBuilder::Target bt = bb.CreateTarget(0)) {
         CHAINERX_ASSERT(internal::GetArrayBody(indices)->nodes().empty());
         bt.Define([indices, axis_norm, a_shape = a.shape()](BackwardContext& bctx) {
             const Array& gout = *bctx.output_grad();
-            bctx.input_grad() = AddAt(Zeros(a_shape, gout.dtype(), gout.device()), indices, axis_norm, gout);
+            Array gx = Zeros(a_shape, gout.dtype(), gout.device());
+            internal::AddAt(gx, indices, axis_norm, gout, gx);
+            bctx.input_grad() = gx;
```
[the remark this answers is not in the document]
Comment (asi1024) [reply chainerx_cc/chainerx/routines/indexing.cc]: Fixed.

File: chainerx_cc/chainerx/routines/indexing.h [hunk 3]
```diff
@@ -16,13 +19,51 @@ Array At(const Array& a, const std::vector<ArrayIndex>& indices);
...
+
+    virtual void Call(const Array& a, const Array& indices, int8_t axis, const Array& b, const Array& out);
+
+protected:
+    virtual void Impl(const Array& a, const Array& indices, int8_t axis, const Array& b, const Array& out) = 0;
+};
+
+namespace internal {
+
+inline void AddAt(const Array& a, const Array& indices, int8_t axis, const Array& b, const Array& out) {
```
Comment (niboshi) [inline chainerx_cc/chainerx/routines/indexing.h]: This one, too?

Comment (asi1024) [reply chainerx_cc/chainerx/routines/indexing.h]: Fixed.

**Comment by niboshi** [comment]
Jenkins, test this please

**Comment by asi1024** [comment]
Fixed for clang-tidy.

**Comment by niboshi** [comment]
Jenkins, test this please

**Comment by niboshi** [comment]
LGTM

**Pull request merged**